### ISSUE TITLE
Allow TestDescriptor to state it will eventually register dynamic tests

### ISSUE THREAD

**Issue by sbrannen** [junit-team/junit5#508]
## Overview

See the discussion in #492 for details.

## Related Issues

- #367
- #407 
- #492 
- #732 

## Deliverables
- [x] Investigate whether we can get rid of `isTest()` and `isContainer()`.
- [ ]  Introduce support for allowing a `TestDescriptor` to state that it will _eventually_ register dynamic tests, if still desirable.

## Initial Team Brainstorming

- For reporting, we somehow need to figure out whether or not a `TestDescriptor` points to a test. Otherwise, we won't be able to tell how many tests have been executed. Therefore, it might not be possible to get rid of `isTest()` or at least the concept of a "test".
- Allowing the dynamic registration of `TestDescriptor`s at any time (also within a test itself) makes the concept of a "container" superfluous. Every `TestDescriptor` may possibly be a container or at least become one (see also #407). Therefore, removing `isContainer()` seems to be reasonable.

_Concerning eventually registering dynamic tests:_

- While we might spawn new tests during runtime, it may be difficult to state whether or not additional tests may be registered. Only after test execution of the current `TestDescriptor` can we give a guaranteed answer that no or all children have been found.
- This is closely related to the extension mechanism and its implementation by a `TestEngine`. One might expect that a `TestDescriptor` `CHILD` below a `TestDescriptor` `PARENT` would inherit the context and state of `PARENT`, i.e. all `BeforeAll/Each` and `AfterAll/Each` statements will get executed. This might be difficult to guarantee (at least for the Jupiter Engine).
  - Having said that, a `hasBeenExecuted()` method might be a good alternative, but it doesn't help us with pruning the `TestPlan` up-front.

**Comment by sbrannen** [comment]
Currently slated for M3 but might get pushed to M4.

**Comment by bechte** [comment]
I will work on a proposal for this.

**Comment by bechte** [comment]
See also: #407 and #520

**Comment by marcphilipp** [comment]
Team decision:

- Remove `isContainer()` from `TestDescriptor` and `TestIdentifier` (maybe deprecate it as a first step).
- Add `willRegisterTests()` (or a better name) to `TestDescriptor` and `TestIdentifier`.
- Remove `hasTests()` from `TestDescriptor` (deprecate first)
- Change pruning to use `isTest() || willRegisterTests()`.

**Comment by gaganis** [comment]
This not being _in progress_ I started looking into it. I have done an initial investigation on how to proceed with this and I have come up with these points and questions.

The to be deprecated methods are used in quite a few places and we will have to express the calling code's intent without using these methods.

From my study of this issue and its related ones, I came to these points on how to do the expression. I would appreciate your opinion and comments on them.

* Do not use the notion of container to distinguish `Identifiers, Descriptors`.
  * This then begs the question how to express eg `jupiter.engine.DisabledTests` 
   which uses `ExecutionEventRecorder::getContainerSkippedCount` or tests using
  `platform.engine.test.event.ExecutionEventConditions::container`

* When we need to disregard Descriptors or Identifiers that in essense do no 
  represent any testing the we will use the `isTest() || [will]RegistersTests()`

Not related with how to express the callers intent, these also require consideration:
* From looking into the code I think `willRegisterTests` is not a good name 
  because it implies that this `Identifier, Descriptor` will do the registration
  in the future but we do need to check this even after we have registered tests
  so this could become misleading. `registersTests()` seems better to me
* This leads in many cases to simply replace the `isContainer()` with 
  `registersTests()` which still implies the notion/concept of a container.

**Comment by sormuras** [comment]
First, thank you @gaganis for your inspection, thoughts and summary regarding this refactoring. Especially, your last bullet point hits the nail on its head: the concept if having containers and tests is valid and will remain.

I'm preparing a PR at the moment, that takes your considertions into account and tries to tackle both:

- is test = is not container ... an identifier/descriptor is a test or it is a container; never both or none.
- demote (hide) TestDescriptor.hasTests() from API

**Comment by sbrannen** [comment]
Hi guys,

I have not looked into PR #720 , but... I _can_ shed some light on the original concepts and rationale in play here.

> is test = is not container ... an identifier/descriptor is a test or it is a container; never both or none.

For starters, I do not agree with the above assumption that "an identifier/descriptor is never both a test and a container".

Here are some things to keep in mind:

1. Do **not** look at JUnit Jupiter when considering whether it makes sense for a something to be both a test and a container.
    - JUnit Jupiter is merely _one_ implementation of a `TestEngine`: it is **not** the guiding light for how all test engines must work.
1. Consider that a different testing framework running on the JUnit Platform might have a need for a node in the test tree that is simultaneously a test **and** a container.
    - In fact, I believe that the Specsy framework does exactly that. If it's not Specsy, it's something else I have seen.

In summary, for this particular feature we have to focus on the _big picture_, the JUnit Platform. 

With that in mind, I believe we need the following possible states for a node -- where being neither a container nor a test is not an option.

1. is a container
1. is a test
1. is a container and a test

Perhaps an enum would best address this topic.

**Comment by sbrannen** [comment]
The issue related to Specsy is #407.

**Comment by sbrannen** [comment]
In my defense, it's been a long time since I thought about any of this in detail, and after rereading #407, I concede that @orfjackal raised some good points.

Maybe it's easier if every node is conceptually a container (without it being explicitly stated), where some nodes _are_ tests, some _contain_ tests, and some are both tests _and_ contain tests.

**Comment by sormuras** [comment]
> Maybe it's easier if every node is conceptually a container (without it being explicitly stated), where some nodes are tests, some contain tests, and some are both tests and contain tests.

Expressed in pseudo-code:
```
Descriptor.isTest() := isEmpty(Descriptor.children())
Descriptor.isContainer() := !Descriptor.isTest()
Descriptor.isTestOrContainer := true
```

**Comment by sormuras** [comment]
The current way #720 is implemented, still allows a tri-state descriptor via overriding the default implementation. Which reads like:
```
	/**
	 * Determine if this descriptor describes a container.
	 *
	 * <p>This default implementation negates the boolean result returned by
	 * {@link #isTest()}. This makes being a test descriptor or a container
	 * descriptor mutual exclusive.
	 *
	 * @return {@code true} if this descriptor is a container; if it is a
	 * test {@code false} is returned.
	 * @see #isTest()
	 */
	default boolean isContainer() {
		return !isTest();
	}
```
If we just change the Javadoc to describe the default implememtation and state _that a different testing framework running on the JUnit Platform might have a need for a node in the test tree that is simultaneously a test and a container._ Everythings fine?

**Comment by gaganis** [comment]
It seems conceptually we're  going in circles here.

> `Descriptor.isTest() := isEmpty(Descriptor.children())`

This statement is not true throughout the entire execution of a test run. At the beginning it might return false and at some later point in time it might. 
And yes... this is what was _originally raised_ by this issue. 

So we again go back to this:
> Maybe it's easier if every node is conceptually a container (without it being explicitly stated), where some nodes are tests, some contain tests, and some are both tests and contain tests.

But at least when I tried to fix this I found that the concept of container was widely used and decision where being made on it's truthfulness, e.g.:

>This then begs the question how to express eg jupiter.engine.DisabledTests
which uses ExecutionEventRecorder::getContainerSkippedCount or tests using
platform.engine.test.event.ExecutionEventConditions::container

If everything is always conceptually a container how will we handle the above?

**Comment by sbrannen** [comment]
Although #720 has been merged into `master`, I am reopening this issue in order to allow us to ensure that we have in fact also addressed the original concern associated with this issue.

**Comment by marcphilipp** [comment]
Team brainstorming ideas:

- Allow engines and containers to be completely dynamic
- Introduce method `mayRegisterTests()` in `TestIdentifier` and `TestDescriptor` for engines/containers to claim that they may register tests
- Remove pruning in Launcher, keep it in Jupiter engine
- Introduce method in `TestPlan` to determine whether it may contain tests and use it in Surefire provider

**Comment by sbrannen** [comment]
> * Allow engines and containers to be completely dynamic

Support for completely dynamic engines was added in #367.

**Comment by sbrannen** [comment]
Introduced _Related Issues_ section.

**Comment by marcphilipp** [comment]
_in progress_
